File: webflux-demo-java/src/main/java/za/co/entelect/springforum/webfluxdemo/dragons/RouteConfiguration.java
package za.co.entelect.springforum.webfluxdemo.dragons;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration
public class RouteConfiguration {

    private final DragonHandler dragonHandler;

    public RouteConfiguration(DragonHandler dragonHandler) {
        this.dragonHandler = dragonHandler;
    }

    @Bean
    public RouterFunction<ServerResponse> dragonRoutes() {
        return route()
                .GET("/dragons",
                        request -> request.queryParam("location").isPresent(),
                        dragonHandler::getDragonsByLocation
                )
                .GET("/dragons", dragonHandler::getAllDragons)
                .GET("/dragons/{name}", dragonHandler::getDragonByName)
                .POST("/dragons", dragonHandler::createDragon)
                .build();
    }
}

File: webflux-demo-kotlin/src/main/kotlin/za/co/entelect/springforum/webfluxkotlin/dragons/RouteConfiguration.kt
package za.co.entelect.springforum.webfluxkotlin.dragons

import org.springframework.web.reactive.function.server.RouterFunctionDsl
import org.springframework.web.reactive.function.server.router

fun dragonRoutes(dragonHandler: DragonHandler) = router {
    "/dragons".nest {
        GET("/{name}") { dragonHandler.getDragonByName(it) }
        (GET("/") and queryParam("location")) { dragonHandler.getDragonsByLocation(it) }
        (GET("/") and !queryParam("location")){ dragonHandler.getAllDragons(it) }
        POST("/") { dragonHandler.createDragon(it) }
    }
}

private fun RouterFunctionDsl.queryParam(name: String) = queryParam(name) { true }


